cfdp-core/src/pdu.rs

```diff
@@ -29,21 +30,19 @@ impl PDUPayload {
         }
     }
 
-    pub fn decode<T: std::io::Read>(
+    pub async fn decode<T: AsyncReadExt + std::marker::Unpin + std::marker::Send>(
```

Comment:
Do you ever expect to read PDUs from a source which does not have all the data at once ?
It seems to me that it complicates the code unnecessarily with async stuff. 

The PDUs are expected to be relatively small in size (couple of KB max) and you read them anyway in memory.

In the default UDP transport - you get a datagram containing one PDU so you have it all in one go. If you implemented a TCP based transport, I would expect you would have some sort of framing decoder which also reads one frame = one PDU at a time. In the space missions I worked with, the PDU is embedded into a TM or TC packet so you process it only when you have the whole packet from the modem.

Making T instead of Read or AsyncRead, be some sort of buffer (tokio Buf?) would be much easier to digest...

Reply:
I see that you also have a UART for Transport.
But don't you need some sort of framing/sync?
How do you know when opening the serial port that you are not starting just in the middle of a PDU?

Reply:
I honestly don't expect that. Originally I didn't want to have to put async on the PDU decoding but ended up needing it for something. It's also the first time I've delved into any of this 😆 so I'm sure I could do better on another pass over it.

Thanks for the suggestion, and actually just looking at this at all.

Regarding the UART, you would probably need additional framing/syncing/etc. I originally added it because I thought it would be convenient for a mission I work on but I'm starting to think maybe it should just be dropped and left implementation specific per user.

Reply:
I haven't looked through your code completely, I rather feel that we needed a bit more refactoring before going to async:
- first I would have eliminated any need for async from the send/recv transaction logic. Those in my mind should be super fast state machines which execute instantly without needing to wait for anything inside ever.
  I know that even after my modification, I pass the transport_tx to the send/recv send_pdu functions and that would imply async - maybe something to reevaluate - pass some sort of closure instead of passing directly the transport_tx? Or instead of calling send_pdu(transport_tx) we could have  transaction.get_next_pdu_to_send() returning the PDU to send.
There is also reading/writing the file data which one could consider for async but from what I read, tokio is not great for asyncing filesystem I/O. So I would keep the reading file data sync for now (as you anyway did I think).

- then as we discussed, I would move away the IPC stuff into its own User defined trait so in the daemon we only deal with channels,  no sockets.

- as already wrote above, PDU decoding in async seems unnecessary complication so I would have changed that to start with a definite buffer of some sort.

After those are done, the async would be limited to:
- reading/writing to the network in the UDP transport
- reading/writing to the IPC in the IpcUser implementation of the User trait
- multiplexing in the daemon between different channels and spawning transactions as async tasks. All the transaction async would be in the daemon spawn_transaction function, not in the actual Send/RecvTransation implementation.

Those are my ideas but feel free to ignore them :)  I'm pretty sure if I actually started implementing them they would probably now work exactly like that.


I haven't done async Rust either and actually not much of Rust at all (I'm in that category who loves the programming language without knowing it :(  ). However I did read an article about the dangers of select!, I think it is this one:
https://blog.yoshuawuyts.com/futures-concurrency-3/

What I understand is that if you use select! and one branch has the PDU async decoding (or async decoding a user primitive), there is a chance that the select! will select the other branch while the PDU on the first branch is only halfway decoded causing  the half decoded PDU to be lost.  
I did't look carefully at your code to see if this can happen but I've seen a select! and remembered about this article so I thought it might be helpful if you weren't' aware of it.

Reply:
I remembered also about this article https://ryhl.io/blog/async-what-is-blocking/  from Alice Ryhl who is on the Tokio team - she writes that one rule of thumb for what blocking means is anything exceeding 100 microsecs.

I did a small test on my refactored sender and it currently spends 64 microsecs in the send_pdu loop - so that's  well in the 100 micros  limit.
If I remove the actual sending from the loop (just returning the result instead of sending) it will go down even more. (I tried just commenting out the transport_tx.send() and it went down to 8 microsecs but probably Rust optimized away a bunch of code)

That is to say that my idea of keeping the main transaction logic sync and putting all the async stuff around it (in the daemon) is feasible.

Reply:
I'm in agreement about the refactoring, I just wanted to try to learn it and then get it recorded so I can refer back to it later. I think in my inexperience there's a lot of things that got changed that don't need to be.
Ideally I'd like only 

 - Communication through channels in the daemon
 - Communication from transports to be async
 - The transaction communication _inside_ the daemon spawn function.
 
 Which I think is inline with your expectations as well.

That's a pretty interesting article (about select). Thanks for the read. I like that the author basically recommends to be smart about spawning tasks as a work around to select anyway.  Like i said, I think inexperience with this kind of code lead to a lot of issues which can be avoided after we've finished implementing some of the refactors ideas and learning more async in general.